apps/andi/lib/andi/input_schemas/cron_tools.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Andi.InputSchemas.CronTools do
+  @moduledoc false
+
+  @more_forgiving_iso_date_format "{YYYY}-{M}-{D}"
+  @more_forgiving_iso_time_format "{h24}:{m}:{s}"
+  @more_forgiving_iso_basic_format "#{@more_forgiving_iso_date_format}T#{@more_forgiving_iso_time_format}"
+
+  def cronstring_to_cronlist!(nil), do: %{}
+  def cronstring_to_cronlist!(""), do: %{}
+  def cronstring_to_cronlist!("never"), do: %{}
+  def cronstring_to_cronlist!("once"), do: %{}
+  def cronstring_to_cronlist!(cronstring) do
+    cronlist = String.split(cronstring, " ")
+    default_keys = [:minute, :hour, :day, :month, :week]
+
+    keys =
+      case crontab_length(cronstring) do
+        6 -> [:second | default_keys]
+        7 -> [:second] ++ default_keys ++ [:year]
+        _ -> default_keys
+      end
+
+    keys
+    |> Enum.zip(cronlist)
+    |> Map.new()
+  end
+
+  def cronlist_to_cronstring!(nil), do: ""
+  def cronlist_to_cronstring!(""), do: ""
+  def cronlist_to_cronstring!(%{"second" => _} = cronlist) do
+    AtomicMap.convert(cronlist, safe: false)
+    |> cronlist_to_cronstring!()
+  end
+  def cronlist_to_cronstring!(%{second: second} = cronlist) when is_nil(second) or second == "" do
+    cronlist
+    |> Map.put(:second, "0")
+    |> cronlist_to_cronstring!()
+  end
+  def cronlist_to_cronstring!(%{second: _second} = cronlist) do
+    [:second, :minute, :hour, :day, :month, :week, :year]
+    |> Enum.reduce("", fn field, acc ->
+      acc <> " " <> to_string(Map.get(cronlist, field, ""))
+    end)
+    |> String.trim_leading()
+    |> String.trim_trailing()
+  end
+  def cronlist_to_cronstring!(cronlist) do
+    cronlist
+    |> Map.put(:second, "0")
+    |> cronlist_to_cronstring!()
+  end
+
+  defp crontab_length(cronstring) do
+    cronstring
+    |> String.split(" ")
+    |> Enum.count()
+  end
+
+  def determine_cadence_type(nil), do: determine_cadence_type("")
+  def determine_cadence_type(cadence) when cadence in ["once", "never"], do: cadence
+  def determine_cadence_type(cadence) do
+    with {:ok, parsed_cadence} <- Crontab.CronExpression.Parser.parse(cadence, true),
+         [_once] <- Crontab.Scheduler.get_next_run_dates(parsed_cadence) |> Enum.take(2) do
```

Comment:
is this enough to check that it's a future ingestion? I might be missing something but can't repeating ingestions still pass this with block